### PR TITLE
executor: solve bug of merge join which does not consume all the inner tuples

### PR DESCRIPTION
**Pull request by fzhedu**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/15301

Problem Summary: In Next(), the merge join does not compeletly consume the inner tuples
It is introduced by https://github.com/pingcap/tidb/pull/14458

### What is changed and how it works?

What's Changed: 
add a loop to comsum all the inner tuples

### Check List 

Tests 

- Unit test

### CODE REVIEW

**Comment by SunRunAway** [comment]
And please add some test cases

File: executor/merge_join.go [hunk 1]
```diff
@@ -354,17 +354,17 @@ func (e *MergeJoinExec) Next(ctx context.Context, req *chunk.Chunk) (err error)
...
-			}
-			e.hasMatch = e.hasMatch || matched
-			e.hasNull = e.hasNull || isNull
-
-			// The inner rows is not exhausted, which means the result chunk is full.
-			// We should keep match context, so return directly.
-			if innerIter.Current() != innerIter.End() && req.IsFull() {
-				return nil
+			// the inner maybe not exhausted at one time
+			for innerIter.Current() != innerIter.End() {
```
[the remark this answers is not in the document]
Comment (fzhedu) [reply executor/merge_join.go]: please note `numToAppend` which limits the matched inners.

[the remark this answers is not in the document]
Comment (fzhedu) [reply executor/merge_join.go]: you can also check the implementation in hash joins.

**Comment by fzhedu** [comment]
> And please add some test cases

add

File: executor/merge_join.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (SunRunAway) [reply executor/merge_join.go]: If `numToAppend` is 0, `chk.IsFull()` would be true.

[the remark this answers is not in the document]
Comment (fzhedu) [reply executor/merge_join.go]: at this case,  `numToAppend` is 1.

File: executor/merge_join_test.go [hunk 2]
```diff
@@ -589,3 +589,30 @@ func (s *testSuiteJoin3) TestVectorizedMergeJoin(c *C) {
...
+	tk.MustExec(`create table Y (a int primary key, b int, index id_b(b));`)
+	tk.MustExec(`insert into Y values (0,2),(2,2);`)
+	tk.MustExec(`create table R (a int primary key, b int);`)
+	tk.MustExec(`insert into R values (2,2);`)
+	tk.MustQuery(`select /*+tidb_smj(R)*/ max(Y.a) from R join Y  on R.a=Y.b where R.b <= Y.a;`).Check(testkit.Rows(
+		`2`,
+	))
+	// the max() limits the required rows at most one
+	// NOTE Y should be the build side
+	tk.MustQuery("explain select /*+tidb_smj(R)*/ max(Y.a) from R join Y  on R.a=Y.b where R.b <= Y.a;").Check(testkit.Rows(
```
Comment (zz-jason) [inline executor/merge_join_test.go]: Please do not compare the execution plan in codes. you can write the test case into a file. If you only want to check whether the query employs the merge join, you can add function like `MustMergeJoin()` to `tk`, we already have `MustIndexLookup()`, `MustPointGet()`, etc.

Comment (fzhedu) [reply executor/merge_join_test.go]: We want to specify the build side as mentioned in the `NOTE`.

**Pull request merged**